### DISCUSSION TITLE
ltree support in queries

### DISCUSSION THREAD

**Discussion by kalmarz** [prest/prest#504] [API Server] [answered]
I'm wondering if I could use `prest` to query in `ltree` fields as documented [here](https://www.postgresql.org/docs/current/ltree.html)?

I have the following:

```
CREATE TABLE tags (
  tag_id INT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  tag ltree NOT NULL,
  scancode VARCHAR(255) NOT NULL,
  CONSTRAINT fk_scancode
    FOREIGN KEY(scancode)
      REFERENCES scans(scancode)
      ON DELETE CASCADE
);

CREATE INDEX tag_gist_index ON tags USING GIST (tag);
```

Querying `ltree` requires different operators, like:

`SELECT * from tags WHERE tag <@ 'BLAH';`

And I couldn't find anything relevant in the documentation, hence asking for the community wisdom.

**Comment by avelino** [comment, marked as answer]
I like the idea of supporting `ltree` field

To support this type of fields would be to implement the list of operators (`Table F.13. ltree Operators`), here https://github.com/prest/prest/blob/9b677c30fef3f06f62f8a22cafd55a075655bdbc/adapters/postgres/postgres.go#L1028:L1067

cc @fabriziomello

**Reply by kalmarz** [reply]
And the support for `ltree` functions would be the icing on the cake.

**Reply by avelino** [reply]
> And the support for `ltree` functions would be the icing on the cake.

@kalmarz the functions in pREST are implemented here: https://github.com/prest/prest/blob/9b677c30fef3f06f62f8a22cafd55a075655bdbc/adapters/postgres/postgres.go#L1293

can you contribute with this implementation? it will be a pleasure to receive your contribution

**Reply by kalmarz** [reply]
I already implemented a few new operators, so yeah... why not? :)

**Reply by fabriziomello** [reply]
Just for future thinking... 

PostgreSQL is a catalog-based database engine, so all operators are stored into a relation named [`pg_catalog.pg_operator`](https://www.postgresql.org/docs/current/catalog-pg-operator.html), so based on that we can generalize this usage and maybe dynamically provide operators by REST api... I really don't know now how difficult can be implement it here but doing that we'll cover all the scenarios, even for custom operators created by [`CREATE OPERATOR`](https://www.postgresql.org/docs/current/sql-createoperator.html) statement.

The query bellow show all operators that the result is a boolean:

```sql
SELECT
    oprleft::regtype, oprname, oprright::regtype
FROM
    pg_operator
WHERE
    oprresult::regtype::text = 'boolean';
```

**Reply by avelino** [reply]
> I already implemented a few new operators, so yeah... why not? :)

@kalmarz did you manage to understand prest's code? do you need help to follow up with the implementation of new operators?

**Reply by avelino** [reply]
> dynamically provide operators

@fabriziomello I'm afraid of having _dynamic_ implementations and leaving **security** holes open for **SQL Injection**

**Reply by fabriziomello** [reply]
> @fabriziomello I'm afraid of having _dynamic_ implementations and leaving **security** holes open for **SQL Injection**

Totally agree with you concern, but what I meant about `dynamically provide operators` is to check if the operator exists in PostgreSQL catalog instead of hard code it ... maybe creating a static configuration to map the pREST operator name to a Postgres operator name so we can load it to pREST memory during the startup or something like that.

Anyway we should somehow create a pREST operator name mapping to Postgres operator name (`pg_catalog.pg_operator.oprname`) without need to change pREST code.
